### PR TITLE
PERF: correctly report memory used by Index's

### PR DESCRIPTION
**Pull request by jreback**

### CODE REVIEW

File: pandas/src/hashtable_class_helper.pxi.in [hunk 1]
```diff
@@ -237,6 +238,12 @@ cdef class {{name}}HashTable(HashTable):
         k = kh_get_{{dtype}}(self.table, key)
         return k != self.table.n_buckets
 
+    def sizeof(self, deep=False):
+        """ return the size of my table in bytes """
+        return self.table.n_buckets * (sizeof(self.table.keys) +
```
Comment (jreback) [inline pandas/src/hashtable_class_helper.pxi.in]: @wesm I believe this is correctly, any thoughts?

**Comment by MaximilianR** [comment]
What do you think about using `__sizeof__` so it works with python?

File: pandas/src/hashtable_class_helper.pxi.in (hunk 1 above)
Comment (chris-b1) [inline pandas/src/hashtable_class_helper.pxi.in]: `sizeof` will return a pointer size here - I think you instead need to do something like this? (not sure the kh aliases are exposed)

```
self.table.n_buckets * (sizeof(khkey_t)  + #keys
                        sizeof({{dtype}}_t)  + #vals
                        sizeof(khint32_t))  #flags
```
                           
https://github.com/pandas-dev/pandas/blob/be328521b3a5525175b08cb740bb970ccf75727e/pandas/src/klib/khash.h#L189

Comment (jreback) [inline pandas/src/hashtable_class_helper.pxi.in]: that makes sense

**Comment by jreback** [comment]
@MaximilianR I added ``__sizeof__``, though you would have to do something like

``sys.getsizeof(index._engine)``, which is simply captured by ``index.memory_usage()`` anyhow

**Pull request closed without merging**